keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1763 @@
+  that reference this ResourceClass. These parameters might describe the kind
+  of resource that are allocated when using the ResourceClass or control
+  cluster-specific options. Because only cluster administrators are allowed to
+  create ResourceClasses and they will be passed separately to the resource
+  driver, they can be used for options that normal users should not be allowed
+  to control.
+
+  For example, a single resource driver might manage different kinds of
+  FPGAs. For each kind, one ResourceClass could define the kind and provide paths
+  to kind-specific tools or URLs for additional resources. Only a handful of
+  ResourceClasses are expected per resource driver.
+
+- ResourceClaim, namespaced, with parameters provided by a normal user
+  that describes a resource instance that needs to be allocated. A
+  ResourceClaim contains the usual metadata, a spec and a status. The
+  spec identifies the driver that handles the resource via a class
+  name.
+
+A third one with information about the driver ("ResourceDriver", similar to the
+"CSIDriver" from storage) could get added in the future. At the moment it is
+not needed yet.
+
+The ResourceClaim spec is immutable. The ResourceClaim
+status is reserved for system usage and holds the current state of the
+resource. The status must not get lost, which in the past was not ruled
+out. For example, status could have been stored in a separate etcd instance
+with lower reliability. To recover after a loss, status was meant to be recoverable.
+A [recent KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/2527-clarify-status-observations-vs-rbac)
+clarified that status will always be stored reliably and can be used as
+proposed in this KEP.
+
+To support arbitrarily complex parameters, both ResourceClass and
+ResourceClaim contain one field which holds a
+[runtime.RawExtension](https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime#RawExtension) (see [user stories](#user-stories) for examples).
+Validation can be handled by drivers through an
+admission controller (if desired) or later at runtime when the
+parameters are passed to the driver.
+
+The reasons for choosing embedded RawExtension fields fields instead of
+references to a custom resource (CR) that is defined by a driver's custom
+resource definition (CRD) are:
+
+- Immutability can be enforced: a separate object could always get
+  deleted and recreated with new content.
+- More user-friendly because there is no need to manage an additional
+  object.
+- Less risky resource driver deployment and updates: CRs get deleted when
+  removing their CRD, which would render existing workloads unusable until
+  users redeploy them.
+
+Kube-scheduler
+and resource driver communicate by modifying that status. The status is also
+how Kubernetes tracks that a driver has allocated the resource and on
+which nodes the resource is available.
+
+This approach is an intentional simplification compared to the
```

Comment:
Personally, while I thought your points about "there's a lot of bugs and races there" is totally valid, it feels like an important abstraction for resources which can be preprovisioned and also take time to preprovision. Maybe I'm missing something obvious.

Reply:
I believe that resource drivers can handle pre-provisioning internally: they can track unused items and use those directly when asked to allocate.

This logic cannot be in Kubernetes because Kubernetes doesn't know whether any existing resource matches the requirements of the resource claim.